File: python/树/106. 从中序与后序遍历序列构造二叉树.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

from typing import List

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if not postorder and not inorder:
            return None
        rootVal = postorder[-1]
        root = TreeNode(rootVal)
        i = inorder.index(rootVal)
        root.left = self.buildTree(inorder[:i],postorder[0:i])
        root.right = self.buildTree(inorder[i+1:],postorder[i:-1])
        return root
    
s = Solution()
inorder = [9,3,15,20,7]
postorder = [9,15,7,20,3]
print(s.buildTree(inorder,postorder))

File: go/tree/106. 从中序与后序遍历序列构造二叉树.go
package tree

func buildTree(inorder []int, postorder []int) *TreeNode {
	if len(inorder) == 0 {
		return nil
	}
	node := &TreeNode{Val: postorder[len(postorder)-1]}
	i := getIndex(inorder, node.Val)
	node.Left = buildTree(inorder[:i], postorder[:i])
	node.Right = buildTree(inorder[i+1:], postorder[i:len(postorder)-1])
	return node
}

func getIndex(A []int, value int) int {
	for i, v := range A {
		if value == v {
			return i
		}
	}
	return 0
}
